sympy/matrices/expressions/matexpr.py

```diff
@@ -589,10 +589,14 @@ class MatrixElement(Expr):
     def __new__(cls, name, n, m):
         n, m = map(_sympify, (n, m))
         from sympy.matrices.matrices import MatrixBase
-        if isinstance(name, (MatrixBase,)):
+        if isinstance(name, MatrixBase):
             if n.is_Integer and m.is_Integer:
                 return name[n, m]
-        if isinstance(name, str):
+            name = _sympify(name)  # change mutable into immutable
+        elif isinstance(name, MatrixSymbol):
+            if not name.valid_index(n, m):
```

Comment:
I think that `valid_index` should always be checked, not just for MatrixSymbol e.g.:
```python
In [18]: n = Symbol('n', integer=True, positive=True)

In [19]: M = MatrixSymbol('M', n, n)

In [20]: M[3, m]
Out[20]: (M)[3, m]

In [21]: M[3, m].subs(M, Matrix([[1, 2], [3, 4]]))
Out[21]: 
⎛⎡1  2⎤⎞      
⎜⎢    ⎥⎟[3, m]
⎝⎣3  4⎦⎠ 
```

Reply:
good point! To me this seems like a strange constructor in that you are creating an unevaluated indexed object rather than simply indexing an object. It works, but having to put these checks here seems odd to me.